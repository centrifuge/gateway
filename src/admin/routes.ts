const index = '/admin';

export default {
  index,
  getAllAccounts: `${index}/accounts`,
  // generateNewAccount: `${index}/accounts/generate`
};