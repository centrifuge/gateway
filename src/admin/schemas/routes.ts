const schemaIndex = '/schemas';

export const schemasRoutes = {
  view: `${schemaIndex}/:id`,
  edit: `${schemaIndex}/:id/edit`,
  index: schemaIndex,
};