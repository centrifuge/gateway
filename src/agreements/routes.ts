const index = '/agreements';

export default {
  index,
};