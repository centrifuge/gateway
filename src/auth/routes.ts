import { ROUTES } from '../common/constants';

const index = '/users';

export default {
  index,
  register: `/register`,
  invite: `${index}/invite`,
  logout: ROUTES.USERS.logout
};
