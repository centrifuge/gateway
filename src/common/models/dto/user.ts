export class User {
  constructor(
    readonly username: string,
    readonly password: string,
    readonly id: string,
  ) {}
}
