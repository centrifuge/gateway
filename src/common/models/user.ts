import { PERMISSIONS } from '../constants';
import { Document, DOCUMENT_ACCESS } from './document';

export interface IUser {
  name: string;
  password?: string;
  email: string;
  _id?: string;
  account: string;
  permissions: PERMISSIONS[];
  schemas?: string[];
  enabled: boolean;
  invited: boolean;
}

export class User implements IUser {
  name: string = '';
  password?: string = '';
  email: string = '';
  _id?: string;
  account: string = '';
  permissions: PERMISSIONS[] = [];
  schemas: string[] = [];
  enabled: boolean;
  invited: boolean;
}

export const canWriteToDoc = (user: User | null, doc?: Document): boolean => {
  if(!user || !doc) return false;
  return accountHasDocAccess(user.account, DOCUMENT_ACCESS.WRITE, doc);
};

export const canReadDoc = (user: User | null, doc?: Document): boolean => {
  if(!user || !doc) return false;
  return accountHasDocAccess(user.account, DOCUMENT_ACCESS.READ, doc);
};

export const accountHasDocAccess = (account: string, access: DOCUMENT_ACCESS, doc?: Document): boolean => {
  return !!(
    doc &&
    doc.header &&
    doc.header[access] &&
    Array.isArray(doc.header[access]) &&
    doc.header[access]!.find(
      ac => ac.toLowerCase() === account.toLowerCase(),
    )
  );
};


export const canCreateDocuments = (user: User): boolean => {
  return (
    user.permissions.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS)
    && user.schemas.length > 0);
};


export const canSignFunding = (user:User | null, doc?:Document):boolean => {
  if(!user) return false;
  return !!(
    doc &&
    doc.attributes &&
    doc.attributes.funding_agreement &&
    Array.isArray(doc.attributes.funding_agreement) &&
    doc.attributes.funding_agreement!.find(
      funding => {
        return funding.funder_id && funding.funder_id.value.toLowerCase() === user.account.toLowerCase()
      },
    )
  );
}
