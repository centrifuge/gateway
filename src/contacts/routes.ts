const index = '/contacts';

export default {
  index,
  new: `${index}/create`,
};
