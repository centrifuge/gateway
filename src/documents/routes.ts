const documentsIndex = '/documents';

export const documentRoutes = {
  index: documentsIndex,
  new: `${documentsIndex}/new`,
  view: `${documentsIndex}/:id`,
  edit: `${documentsIndex}/:id/edit`,
};
