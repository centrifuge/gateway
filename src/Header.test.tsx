import React from 'react';
import Header from './Header';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import getRootReducer from './store/reducers';
import { User } from './common/models/user';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import routes from './routes';
import { AxisTheme } from '@centrifuge/axis-theme';

const store = createStore(getRootReducer({}), { router: { location: { pathname: '/' } } });

describe('Header', () => {

  const user = new User();
  const push = (route: string) => route;
  it('Should render an empty header', () => {
    const bodyShallow = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Header selectedRoute={'/'} menuItems={[]} user={null} push={push}/>
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(bodyShallow).toMatchSnapshot();
  });

  it('Should render the Header', () => {

    const items = [
      { label: 'Invoices', route: routes.invoices.index },
      { label: 'Contacts', route: routes.contacts.index },
      { label: 'Logout', route: routes.user.logout, external: true, secondary: true },
    ];

    const user = new User();
    user.account = '0x33333';

    const bodyShallow = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <AxisTheme>
            <Header selectedRoute={'/'} user={user} menuItems={items} push={push}/>
          </AxisTheme>
        </BrowserRouter>
      </Provider>
  ,
  ).toJSON();
  expect(bodyShallow).toMatchSnapshot();
  });

  });
