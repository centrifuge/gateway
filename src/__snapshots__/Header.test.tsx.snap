// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Should render an empty header 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 iEnTcK"
  style={
    Object {
      "height": "90px",
      "position": "sticky",
      "top": 0,
      "zIndex": 1,
    }
  }
>
  <div
    className="StyledBox-sc-13pk1d4-0 fbWwXK"
  >
    <a
      href="/"
      label="Centrifuge"
      onClick={[Function]}
      size="large"
    >
      <img
        className="StyledImage-ey4zx9-0 fmkmCl"
        src="logo.png"
      />
    </a>
    <div
      className="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo"
    />
    <div
      className="StyledBox-sc-13pk1d4-0 ffeAql"
    />
    <div
      className="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo"
    />
    <div
      className="StyledBox-sc-13pk1d4-0 BclQN"
    />
  </div>
</div>
`;

exports[`Header Should render the Header 1`] = `
<div
  className="StyledGrommet-sc-19lkkz7-0 eWRwfm"
>
  <div
    className="StyledBox-sc-13pk1d4-0 gkxoZn"
    style={
      Object {
        "height": "90px",
        "position": "sticky",
        "top": 0,
        "zIndex": 1,
      }
    }
  >
    <div
      className="StyledBox-sc-13pk1d4-0 fbWwXK"
    >
      <a
        href="/"
        label="Centrifuge"
        onClick={[Function]}
        size="large"
      >
        <img
          className="StyledImage-ey4zx9-0 fmkmCl"
          src="logo.png"
        />
      </a>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo"
      />
      <div
        className="StyledBox-sc-13pk1d4-0 ffeAql"
      >
        <button
          className="StyledButton-sc-323bzc-0 kPUXNH"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          Invoices
        </button>
        <div
          className="StyledBox__StyledBoxGap-sc-13pk1d4-1 iChEkS"
        />
        <button
          className="StyledButton-sc-323bzc-0 kPUXNH"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          Contacts
        </button>
      </div>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo"
      />
      <div
        className="StyledBox-sc-13pk1d4-0 BclQN"
      >
        <div
          className="StyledBox-sc-13pk1d4-0 gQLnBt"
        >
          <span
            className="StyledText-sc-1sadyjn-0 cTRjlw"
          >
            Centrifuge ID: 
          </span>
          <div
            className="StyledBox__StyledBoxGap-sc-13pk1d4-1 gufrXq"
          />
          <div
            className="StyledBox-sc-13pk1d4-0 cMsezL"
          >
            <div
              className="DisplayField__DisplayFieldBox-lfek92-0 bzzYlY StyledBox-sc-13pk1d4-0 IRSNj"
            >
              <p
                className="StyledParagraph-tbetod-0 giEJAX"
              >
                0x33333
              </p>
            </div>
          </div>
        </div>
        <div
          className="StyledBox__StyledBoxGap-sc-13pk1d4-1 iChEkS"
        />
        <span
          className="StyledText-sc-1sadyjn-0 cTRjlw"
        >
           
        </span>
      </div>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo"
      />
      <div
        className="StyledBox-sc-13pk1d4-0 BclQN"
      >
        <a
          className="StyledAnchor-sc-1rp7lwl-0 cWvlCm"
          href="/api/users/logout"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Logout
        </a>
      </div>
    </div>
  </div>
</div>
`;
